### PR TITLE
[GEOS-10694] Suppress stacktrace on startup when logging.xml is missing.

### PR DESCRIPTION
**Pull request by dieterstueken**
[GEOS-10685](https://osgeo-org.atlassian.net/browse/GEOS-10685)

[GEOS-10694](https://osgeo-org.atlassian.net/browse/GEOS-10694)
[GEOS-10693](https://osgeo-org.atlassian.net/browse/GEOS-10693)

Starting on an empty context without any logging.xml generates an error. 
Even worse: an empty logging.xml file is generated instead. This raises future future errors.
(The possibly deeper problem here is, that `Resource.in()` creates an empty new file.)

Expected behavior: silently ignore the empty logging.xml and don't create an empty one.

Committed a unit test to demonstrate the error an a fix to resolve it.

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [ ] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

**Comment by aaime** [comment]
I don't think this fix matches the description of https://osgeo-org.atlassian.net/browse/GEOS-10685
The ticket must describe the issue you found, if there is no ticket matching it, you should create a new one.

@jodygarnett this area of code is something you worked on recently I believe, could you have a look at the changes?

**Comment by dieterstueken** [comment]
This tries to solve the first topic of [GEOS-10685](https://osgeo-org.atlassian.net/browse/GEOS-10685):

> 1. Usual message when trying to configure without logging.xml

I may split it off into a separate ticket, but that does not contain any additional information.

**Comment by aaime** [comment]
@dieterstueken we want tickets specifically describing the change, not something related. When this change is committed, the ticket is closed and gets into the release notes. With a vague relation, we either close the ticket, and the release notes are incorrect about the scope of changes, or we don't, and the change is not reported to end users.

File: src/main/src/main/java/org/geoserver/logging/LoggingStartupContextListener.java [hunk 1]
```diff
@@ -112,7 +112,7 @@ public void contextInitialized(ServletContextEvent event) {
         // the "change" between logging.xml and the versions stored in JDBC. KS
         // TODO find a better solution than re-initializing on JDBCCOnfig startup.
         Resource f = store.get("logging.xml");
-        if (f != null) {
+        if (f != null && f.getType() == Resource.Type.RESOURCE) {
```
Comment (jodygarnett) [inline src/main/src/main/java/org/geoserver/logging/LoggingStartupContextListener.java:115]: Nice minimal change! Although I am not sure `f` can ever be ``null``.

**Review by jodygarnett — APPROVED** [review]
Nice minimal change; suitable for backport

**Comment by jodygarnett** [comment]
> @dieterstueken we want tickets specifically describing the change, not something related. When this change is committed, the ticket is closed and gets into the release notes. With a vague relation, we either close the ticket, and the release notes are incorrect about the scope of changes, or we don't, and the change is not reported to end users.

Andrea in this case I reported the ticket, we can make "sub tasks": 

* https://osgeo-org.atlassian.net/browse/GEOS-10694 Creating a logging.xml file when starting up form an empty directory is expected and should not log a stacktrace

File: src/main/src/test/java/org/geoserver/logging/LoggingStartupContextListenerTest.java [hunk 2]
```diff
@@ -127,6 +128,26 @@ public void testLogLocationFromServletContext() throws Exception {
...
+        assertTrue(logs.mkdirs());
+
+        MockServletContext context = new MockServletContext();
+        context.setInitParameter("GEOSERVER_DATA_DIR", tmp.getPath());
+        context.setInitParameter(
+                "GEOSERVER_LOG_LOCATION", new File(tmp, "foo.log").getAbsolutePath());
+
+        ServletContextListener listener = new LoggingStartupContextListener();
+
+        listener.contextInitialized(new ServletContextEvent(context));
```
Comment (jodygarnett) [inline src/main/src/test/java/org/geoserver/logging/LoggingStartupContextListenerTest.java:147]: The test case does not seem to listen to the logs to see if an logging event is generated. I made an test appender for writing these kind of test cases; but this is a bit of a special case since there is not yet an existing logging setup to add a test appender to! 

So this test case is good and we should keep it; but it does not check that logging messages for the log file being missing are indeed suppressed.

Comment (aaime) [reply src/main/src/test/java/org/geoserver/logging/LoggingStartupContextListenerTest.java:147]: @dieterstueken can you add the missing tests?

Comment (dieterstueken) [reply src/main/src/test/java/org/geoserver/logging/LoggingStartupContextListenerTest.java:147]: will have a look this evening ...

Comment (dieterstueken) [reply src/main/src/test/java/org/geoserver/logging/LoggingStartupContextListenerTest.java:147]: My initial intention was just to avoid the NPE.
(sorry for the different issues flipping back and forth, 
something with the badge processing i did not get right)

Without the fix the NPE is raised and after the fix it vanished. So, the test works fine for me.

That the logging generation must be suppressed if the config file is missing, is an additional requirement i was not aware of. But I may simply add a test if the log directory really remains empty after all.

I also observed that the test does not finally remove the directories created (but this is true for all other tests, too).

Comment (aaime) [reply src/main/src/test/java/org/geoserver/logging/LoggingStartupContextListenerTest.java:147]: Not removing the directories created is an intentional decision, allows the developer to inspect the data dir contents used by the tests. Very important when pulling hairs debugging a complex situation. The directories are placed in ``/target``, which is meant to contain removable information anyways (so ``mvn clean`` will get rid of them).

Comment (jodygarnett) [reply src/main/src/test/java/org/geoserver/logging/LoggingStartupContextListenerTest.java:147]: Here is the code snippet - however TestAppender was not collecting events when I tried:

```
try (TestAppender appender = TestAppender.createAppender("quite", null)) {
            appender.startRecording("org.geoserver.logging");
            
            appender.trigger("Could not reconfigure LOG4J loggers");
            
            LoggingStartupContextListener listener = new LoggingStartupContextListener();
            listener.contextInitialized(new ServletContextEvent(context));
            listener.contextDestroyed(new ServletContextEvent(context));

            appender.stopRecording("org.geoserver.logging");
        }
```

The code snippet is correct though, and when TestAppender functionality is fixed it will be correct.

**Comment by jodygarnett** [comment]
When this is done building it can be squashed and merged
